File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.js
function solution(n, wires) {
    let answer = 100;
    
    for(let i = 0; i <= n; i++) {
        map.set(i, []);
        visited.push(false);
    }
    
    for(let i = 0; i < wires.length; i++) {
        const x = wires[i][0];
        const y = wires[i][1];
        
        let arr1 = map.get(x);
        arr1.push(y);
        map.set(x, arr1);
        
        let arr2 = map.get(y);
        arr2.push(x);
        map.set(y, arr2);
    }
    
    
    for(let i = 0; i < wires.length; i++) {
        const x = wires[i][0];
        const y = wires[i][1];
        
        resetFilter(n);
        
        dfs(x, y);
        const n1 = cnt;
        
        resetFilter(n);
        
        dfs(y, x);
        const n2 = cnt;
        
        answer = Math.min(answer, Math.abs(n1 - n2));
    }
    
    return answer;
}

let map = new Map();
let visited = [];
let cnt = 0;

const resetFilter = (n) => {
    cnt = 0;
    for(let i = 0; i <= n; i++) {
        visited[i] = false;
    }
}

const dfs = (a, b) => {
    visited[a] = true;
    cnt++;
    
    for(let n of map.get(a)) {
        if(n === b) continue;
        if(visited[n]) continue;
        
        dfs(n, b);
    }
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    
    private static int length, count;
    private static boolean[] visit;
    private static boolean[][] map;
    
    public int solution(int n, int[][] wires) {
        int answer = 100;
        length = n;
        map = new boolean[n + 1][n + 1];
        
        for(int[] wire : wires) {
            int v1 = wire[0];
            int v2 = wire[1];
            
            map[v1][v2] = true;
            map[v2][v1] = true;
        }

        for(int[] wire : wires) {
            visit = new boolean[n + 1];
            
            int v1 = wire[0];
            int v2 = wire[1];
            
            map[v1][v2] = false;
            map[v2][v1] = false;
            
            count = 0;
            search(v1);
            int a = count;
            
            count = 0;
            search(v2);
            int b = count;
            
            int diff = Math.abs(a - b);
            
            answer = Math.min(answer, diff);
            
            map[v1][v2] = true;
            map[v2][v1] = true;
        }
        
        return answer;
    }
    
    private static void search(int x) {
        visit[x] = true;
        count++;
        
        for(int i = 1; i <= length; i++) {
            if(visit[i] || x == i) continue;
            if(map[x][i] == true) {
                search(i);
            }
        }
    }
}